src/VisualStudio/Core/Def/Implementation/Library/FindResults/LibraryManager_FindReferences.cs

```diff
@@ -50,30 +51,59 @@ private IList<AbstractListItem> CreateFindReferencesItems(Solution solution, IEn
 
                 var definition = referencedSymbol.Definition;
                 var locations = definition.Locations;
-                uniqueLocations = AddLocations(solution, list, uniqueLocations, locations, definition.GetGlyph());
-                uniqueLocations = AddLocations(solution, list, uniqueLocations, referencedSymbol.Locations.Select(loc => loc.Location), Glyph.Reference);
 
-                string filePath;
-                int lineNumber, charOffset;
-                var symbolNavigationService = solution.Workspace.Services.GetService<ISymbolNavigationService>();
+                foreach (var definitionLocation in definition.Locations)
+                {
+                    var definitionItem = CreateDefinitionItem(solution, referencedSymbol, uniqueLocations, definitionLocation, definition.GetGlyph());
+                    if (definitionItem != null)
+                    {
+                        definitions.Add(definitionItem);
+                        definitionItem.Children.AddRange(
+                            CreateReferenceItems(solution, uniqueLocations, referencedSymbol.Locations.Select(loc => loc.Location), Glyph.Reference));
+                    }
+                }
 
                 // Add on any definition locations from third party language services
+                string filePath;
+                int lineNumber, charOffset;                
                 if (symbolNavigationService.WouldNavigateToSymbol(definition, solution, out filePath, out lineNumber, out charOffset))
                 {
-                    list.Add(new ExternalListItem(filePath, lineNumber, charOffset, definition.Name, definition.GetGlyph().GetGlyphIndex(), this.ServiceProvider));
+                    definitions.Add(new ExternalTreeItem(filePath, lineNumber, charOffset, definition.Name, definition.GetGlyph().GetGlyphIndex(), this.ServiceProvider));
                 }
             }
 
-            return list;
+            return definitions;
         }
 
-        private HashSet<ValueTuple<Document, TextSpan>> AddLocations(
+        private AbstractTreeItem CreateDefinitionItem(
             Solution solution,
-            List<AbstractListItem> list,
+            ReferencedSymbol referencedSymbol,
             HashSet<ValueTuple<Document, TextSpan>> uniqueLocations,
-            IEnumerable<Location> locations,
+            Location location,
             Glyph glyph)
         {
+            if (!location.IsInSource)
+            {
+                return referencedSymbol.Locations.Any()
+                    ? new MetadataDefinitionTreeItem(referencedSymbol.Definition, glyph.GetGlyphIndex())
```

Comment:
Do we get references to hidden locations as well or are they already filtered out?